### PR TITLE
[move-compiler] Added a linter flagging potential unenforceable custom transfer/share/freeze policy

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This adds another linter, this time to flag a potential unenforceable transfer/share/freeze policy - a developer may implement a custom policy for objects with the `store` ability which can be transferred/shared/frozen using public function variants.

## Test Plan 

New tests have been added.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

When building Move code, additional linter warnings related to implementing custom transfer/share/freeze functions may appear. These functions are created to enforce a custom transfer/share/freeze policy but implementing them to work with instances of  a type with the `store` ability makes the policy unenforceable (these type instances can be transferred/shared/frozen using public variants of transfer/share/freeze functions).

### CODE REVIEW

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                let action = if *fname == TRANSFER_FUN {
+                    "transferred"
+                } else if *fname == SHARE_FUN {
+                    "shared"
+                } else {
+                    "frozen"
+                };
+                let uid_msg = format!("Instances of a type with a store ability can be {} using \
+                                       public_{} function which often makes calling the (private) {} \
+                                       function variant in the module defining this type unnecessary", action, fname, fname);
```
Comment (tnowacki) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:162]: Maybe we should change the wording here. To me its less that these are unnecessary, but rather dangerous? Since you cannot guarantee that your custom transfer function will be called.

**Review by tnowacki — COMMENTED** [review]
Overall looks good! Just some thoughts on the wording

File: crates/sui-move-build/src/linters/redundant_custom_call.rs (hunk 1 above)
Comment (awelc) [reply crates/sui-move-build/src/linters/redundant_custom_call.rs]: Good point. Please take a look at the new take.

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+};
+use std::collections::BTreeMap;
+
+use super::INVALID_LOC;
+
+const TRANSFER_FUN: &str = "transfer";
+const SHARE_FUN: &str = "share_object";
+const FREEZE_FUN: &str = "freeze_object";
+
+const PRIVATE_OBJ_FUNCTIONS: &[(&str, &str, &str)] = &[
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:42]: Not for this PR, but should we have a central place where all of these functions (and other special names) are defined? e.g., the sui-verifier also uses some of these names, and it feels like a bit of a code smell to have them sprinkled about the codebase instead of being defined in one central module.

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                    ("transfer", "transferred")
+                } else if *fname == SHARE_FUN {
+                    ("share", "shared")
+                } else {
+                    ("freeze", "frozen")
+                };
+                let uid_msg = format!("Instances of a type with a store ability can be {} using the \
+                                       public_{} function which often negates the intent of enforcing \
+                                       custom {} policy (implemented through calling the private {} \
+                                       function variant in the module defining this type)", action, fname, op, fname);
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:164]: I'm a big fan of inline usage of variables in string formatting where possible. So thoughts on this? It's a personal preference, so feel free to not apply though if you prefer the current way.

```suggestion
                let uid_msg = format!("Instances of a type with a store ability can be {action} using the \
                                       public_{fname} function which often negates the intent of enforcing \
                                       a custom {op} policy (implemented through calling the private {fname} \
                                       function variant in the module defining this type)");
```

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        {
+            if let Value::LocalObjWithStore(obj_addr_loc) = args[0] {
+                let msg = format!(
+                    "Potential unintended implementation of a custom {} function.",
+                    fname
+                );
+                let (op, action) = if *fname == TRANSFER_FUN {
+                    ("transfer", "transferred")
+                } else if *fname == SHARE_FUN {
+                    ("share", "shared")
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:157]: Nit: should this be "sharing"?

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                } else {
+                    ("freeze", "frozen")
+                };
+                let uid_msg = format!("Instances of a type with a store ability can be {} using the \
+                                       public_{} function which often negates the intent of enforcing \
+                                       custom {} policy (implemented through calling the private {} \
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:163]: Nit: "enforcing \ custom .." => "enforcing \ a custom .."

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        _state: &mut State,
+        _l: &LValue,
+        _value: &Value,
+    ) -> bool {
+        false
+    }
+}
+
+fn is_local_obj_with_store(st: &SingleType, context: &CFGContext) -> bool {
+    let sp!(_, st_) = st;
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:200]: mega nit: you can directly deconstruct this in the function declaration:
```rust
fn is_local_obj_with_store(sp!(_, st_): &SingleType, ...)
```

File: crates/sui-move-build/src/linters/redundant_custom_call.rs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    let bt = match st_ {
+        SingleType_::Base(v) => v,
+        // transfer/share/freeze take objects by value so even if by-reference object has store and
+        // is module-local, it could not end up being an argument to one of these functions
+        SingleType_::Ref(_, _) => return false,
+    };
+    let sp!(_, bt_) = bt;
```
Comment (tzakian) [inline crates/sui-move-build/src/linters/redundant_custom_call.rs:207]: Nit: same here for the deconstruction on `bt` -- this deconstruction can just be lifted to the binding of `bt` I believe?

File: crates/sui-move-build/tests/linter/redundant_custom.exp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+warning[Lint W01001]: potentially unenforceable custom transfer/share/freeze policy
+   ┌─ tests/linter/redundant_custom.move:21:16
+   │
+21 │     public fun custom_freeze_bad(o: S1) {
+   │                ^^^^^^^^^^^^^^^^^ - An instance of a module-private type with a store ability to be frozen coming from here
+   │                │                  
+   │                Potential unintended implementation of a custom freeze_object function.
+22 │         transfer::freeze_object(o)
+   │                   ------------- Instances of a type with a store ability can be frozen using the public_freeze_object function which often negates the intent of enforcing custom freeze policy (implemented through calling the private freeze_object function variant in the module defining this type)
```
Comment (tzakian) [inline crates/sui-move-build/tests/linter/redundant_custom.exp:51]: The length of this message makes me wonder if we should move this into a help message on this warning, and have a shorter error message for this?

**Review by tzakian — APPROVED** [review]
Overall looks good to me! No real blockers. However I did leave a bunch of nits and a couple more philosophical questions worth pondering over that aren't directly related to this PR.

File: crates/sui-move-build/src/linters/redundant_custom_call.rs (hunk 3 above)
Comment (awelc) [reply crates/sui-move-build/src/linters/redundant_custom_call.rs:164]: I don't have a strong preference so changed it accordingly

File: crates/sui-move-build/tests/linter/redundant_custom.exp (hunk 8 above)
Comment (awelc) [reply crates/sui-move-build/tests/linter/redundant_custom.exp:51]: I will leave it up to @tnowacki to decide. It's a tough tradeoff here between the succinctness and the desire for the message to be descriptive enough...

Comment (tzakian) [reply crates/sui-move-build/tests/linter/redundant_custom.exp:51]: Totally agreed. It's a fine line (and actually incredibly difficult IMO) to determine these types of things. 

My two cents are this should most likely become a help message eventually, BUT if we don't already really use "help" in our error messages elsewhere it's probably best to not start now and keep this as-is -- and make adding in help messages and the like a separate bit of work.

Comment (tnowacki) [reply crates/sui-move-build/tests/linter/redundant_custom.exp:51]: We do have a "note" section in diagnostics. So it might be worthwhile to move a lot of that text into a note.

Comment (tnowacki) [reply crates/sui-move-build/tests/linter/redundant_custom.exp:51]: I think we don't quite have any block of text quite this large in other diagnostics. I have used a note for some edge case / ambiguous behavior for local variable copy/move. 
This feels appropriate here too (maybe)

Comment (awelc) [reply crates/sui-move-build/tests/linter/redundant_custom.exp]: Gave it a shot!

File: crates/sui-move-build/tests/linter/redundant_custom.exp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+warning[Lint W01001]: potentially unenforceable custom transfer/share/freeze policy
+   ┌─ tests/linter/redundant_custom.move:21:16
+   │
+21 │     public fun custom_freeze_bad(o: S1) {
+   │                ^^^^^^^^^^^^^^^^^ - An instance of a module-private type with a store ability to be frozen coming from here
+   │                │                  
+   │                Potential unintended implementation of a custom freeze_object function.
+22 │         transfer::freeze_object(o)
+   │                   ------------- Instances of a type with a store ability can be frozen using the public_freeze_object function which often negates the intent of enforcing a custom freeze policy
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/redundant_custom.exp:55]: Still might be a tad on the long side, but the split is nice :)

**Review by tnowacki — APPROVED** [review]
@tzakian already approved, but just giving it a second stamp. Good for me :)
